ctapipe/tools/make_irf.py

```diff
@@ -0,0 +1,353 @@
+                reduced_events["proton_count"],
+                reduced_events["electron_count"],
+            )
+        )
+
+        self.signal_events = reduced_events["gamma"]
+        self.background_events = vstack(
+            [reduced_events["proton"], reduced_events["electron"]]
+        )
+        self.signal_events["selected_gh"] = evaluate_binned_cut(
+            self.signal_events["gh_score"],
+            self.signal_events["reco_energy"],
+            self.opt_result.gh_cuts,
+            operator.ge,
+        )
+        self.background_events["selected_gh"] = evaluate_binned_cut(
+            self.background_events["gh_score"],
+            self.background_events["reco_energy"],
+            self.opt_result.gh_cuts,
+            operator.ge,
+        )
+        self.theta_cuts_opt = self.theta.calculate_theta_cuts(
+            self.signal_events[self.signal_events["selected_gh"]]["theta"],
+            self.signal_events[self.signal_events["selected_gh"]]["reco_energy"],
+            self.reco_energy_bins(),
+        )
+
+        self.signal_events["selected_theta"] = evaluate_binned_cut(
+            self.signal_events["theta"],
+            self.signal_events["reco_energy"],
+            self.theta_cuts_opt,
+            operator.le,
+        )
+        self.signal_events["selected"] = (
+            self.signal_events["selected_theta"] & self.signal_events["selected_gh"]
+        )
+
+        self.background_events["selected_theta"] = evaluate_binned_cut(
+            self.background_events["theta"],
+            self.background_events["reco_energy"],
+            self.theta_cuts_opt,
+            operator.le,
+        )
+
+        # TODO: rework the above so we can give the number per
+        # species
+        self.log.debug(
+            "Keeping %d signal, %d backgrond events"
+            % (
+                sum(self.signal_events["selected"]),
+                sum(self.background_events["selected"]),
+            )
+        )
+
+        # calculate sensitivity
+        signal_hist = create_histogram_table(
+            self.signal_events[self.signal_events["selected"]],
+            bins=self.reco_energy_bins,
+        )
+
+        background_hist = estimate_background(
+            self.background_events[self.background_events["selected_gh"]],
+            reco_energy_bins=self.reco_energy_bins,
+            theta_cuts=self.theta_cuts_opt,
+            alpha=self.alpha,
+            fov_offset_min=self.bins.fov_offset_min,
+            fov_offset_max=self.bins.fov_offset_max,
+        )
+        self.sensitivity = calculate_sensitivity(
+            signal_hist, background_hist, alpha=self.alpha
+        )
+
+        # scale relative sensitivity by Crab flux to get the flux sensitivity
+        self.sensitivity["flux_sensitivity"] = self.sensitivity[
+            "relative_sensitivity"
+        ] * self.spectrum(self.sensitivity["reco_energy_center"])
+
+    def finish(self):
```

Comment:
It seems weird to me to do the actual IRF computations in `finish`, I'd move them to the main `run` method.